cocos/core/components/base-renderer.ts

```diff
@@ -0,0 +1,220 @@
+/*
+ Copyright (c) 2020 Xiamen Yaji Software Co., Ltd.
+
+ https://www.cocos.com/
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+/**
+ * @packageDocumentation
+ * @module model
+ */
+
+import { EDITOR } from 'internal:constants';
+import {
+    ccclass, type, displayOrder, displayName, serializable,
+} from 'cc.decorator';
+import { Material } from '../assets/material';
+import { Component } from './component';
+import { IMaterialInstanceInfo, MaterialInstance } from '../renderer/core/material-instance';
+import { warnID } from '../platform/debug';
+import { legacyCC } from '../global-exports';
+
+const _matInsInfo: IMaterialInstanceInfo = {
+    parent: null!,
+    owner: null!,
+    subModelIdx: 0,
+};
+
+@ccclass('cc.BaseRenderer')
+export class BaseRenderer extends Component {
+    @type(Material)
+    @displayOrder(0)
+    @displayName('Materials')
+    get sharedMaterials () {
+        // if we don't create an array copy, the editor will modify the original array directly.
+        return (EDITOR && this._materials.slice()) || this._materials;
+    }
+
+    set sharedMaterials (val) {
+        for (let i = 0; i < val.length; i++) {
```

Comment:
I found much codes caches length like this
```javascript
for (let i = 0, len = val.length; i < len; ++)
```
Does caching length will improve performance?